### ISSUE TITLE
Incomplete resolving of subclass chains in `runtime-evaluated-base-classes`

### ISSUE THREAD

**Issue by disrupted** [astral-sh/ruff#7893]
Extracted from https://github.com/astral-sh/ruff/issues/7866#issuecomment-1752954586

Ruff doesn't reliably detect the base class (even within the same file).

given this example
```python
from abc import ABC
from pydantic import BaseModel

class Model(BaseModel):
    x: str


class Empty(BaseModel):
    ...


class Abstract(Empty, ABC):
    inner: Model
```

Ruff evaluates the following classes and their bases (I simply placed some log statements in this function)
https://github.com/astral-sh/ruff/blob/a1509dfc7c2b16f3762545fc802d49f5f03726e2/crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs#L42

```log
class Model  bases: [Some(["pydantic", "BaseModel"])]
class Abstract  bases: [None, Some(["abc", "ABC"])]
```
notice the absence of `Empty` and how `Abstract` then doesn't get detected as `BaseModel`.

Therefore, `flake8-type-checking` rules often result in false positives as it's unable to correctly resolve the base class.
